stablehlo/dialect/TypeInference.cpp

```diff
@@ -302,6 +303,62 @@ unsigned potentiallyComplexBitwidth(Type type) {
                    : type.getIntOrFloatBitWidth();
 }
 
+LogicalResult verifyReplicaGroups(Optional<Location> location,
+                                  DenseIntElementsAttr attr,
+                                  bool useGlobalDeviceIdsAvailableAndTrue,
+                                  bool isUniformSized,
+                                  Optional<size_t> expectedSubgroupSize) {
+  auto replicaGroupType = attr.getType().dyn_cast<RankedTensorType>();
+  if (!replicaGroupType || replicaGroupType.getRank() != 2 ||
+      !replicaGroupType.getElementType().isInteger(/*width=*/64))
+    return emitOptionalError(
+        location,
+        "replica groups should be a rank 2 tensor of 64 bit integers");
+
+  if (replicaGroupType.getShape()[0] == 0 ||
+      replicaGroupType.getShape()[1] == 0) {
+    if (useGlobalDeviceIdsAvailableAndTrue)
+      return emitOptionalError(
+          location,
+          "if `use_global_device_ids` is set, the replica groups cannot be "
+          "empty");
+    return success();
+  }
+
+  auto ids = attr.getValues<int64_t>();
+  llvm::SmallSet<int64_t, 8> replicaSeen;
+  for (int64_t id : ids) {
+    // Replica groups are stored in a 2D tensor. If the op supports non-uniform
+    // groups, null replica IDs are stored as -1.
+    if (id == -1) {
+      if (isUniformSized) {
+        return emitOptionalError(location, "Invalid replica id -1");
+      }
+      continue;
+    }
+
+    if (!replicaSeen.insert(id).second) {
+      return emitOptionalError(location, "replica id #", id,
+                               " seen more than once");
+    }
```

Comment:
I also wish we could check that these replica ids are consistent with `num_replicas` and `num_partitions`, but that will have to wait until https://github.com/openxla/stablehlo/issues/425#issuecomment-1327785977. Let's open a ticket for this though.

Reply:
Done.
https://github.com/openxla/stablehlo/issues/650